docs/source/gateway.rst

```diff
@@ -0,0 +1,745 @@
+layer of security for API key management, make the `MLflow Gateway` service an ideal choice for
+businesses that utilize language models on a regular basis.
+
+.. contents:: Table of Contents
+  :local:
+  :depth: 2
+
+.. _gateway-concepts:
+
+Concepts
+========
+
+Gateway Configuration
+---------------------
+
+The `MLflow Gateway` service relies on a user-provided configuration file, written in YAML,
+that defines the routes and providers available to the service. The configuration file dictates
+how the gateway interacts with various language model providers and determines the end-points that
+users can access.
+
+The configuration file includes a series of sections, each representing a unique route.
+Each route section has a name, a type, and a model specification, which includes the model
+provider, name, and configuration details. The configuration section typically contains the base
+URL for the API and an environment variable for the API key.
+
+Here is an example of a simple single-route configuration:
+
+.. code-block:: yaml
+
+    routes:
+      - name: chat
+        type: llm/v1/chat
+        model:
+          provider: openai
+          name: gpt-3.5-turbo
+          config:
+            openai_api_base: https://api.openai.com/v1
+            openai_api_key: $OPENAI_API_KEY
+
+
+In this example, we define a route named ``chat`` that corresponds to the ``llm/v1/chat`` type, which
+will use the ``gpt-3.5-turbo`` model from `OpenAI` to return query responses from the `OpenAI` service.
+
+The beauty of the Gateway Configuration lies in its flexibility and ease of use.
+If the configuration file is updated at the target location, the `Mlflow Gateway` service can hot-swap
+routes, allowing new providers or model types to be incorporated without requiring changes to
+the applications interfacing with the gateway. The hot-swap functionality is autonomous and
+requires no server downtime or manual restarting of the service.
+This allows for simple, seamless changes to the underlying models and providers while keeping
+your applications steady and reliable.
+
+In order to define an API KEY for a given provider, there are three primary options:
+
+1. Directly include it in the YAML configuration file.
+2. Use an environment variable to store the API key and reference it in the YAML configuration file.
+3. Define your API key in a file and reference the location of that key-bearing file within the YAML configuration file.
+
+If you choose to include the API key directly, replace ``$OPENAI_API_KEY`` in the YAML file with your
+actual API key.
+
+If you prefer to use an environment variable (recommended), you can define it in your shell
+environment. For example:
+
+.. code-block:: bash
+
+     export OPENAI_API_KEY="your_openai_api_key"
+
+**Note:** Replace "your_openai_api_key" with your actual `OpenAI` API key.
+
+Remember to ensure secure access to the system that the `MLflow Gateway` service is running in to
+protect access to these keys.
```

Comment:
Can we call out a specific recommendation, such as running the gateway behind a reverse proxy authentication layer?